README.md
# @vgrid/react-leaflet-heatmap-layer

`react-leaflet-heatmap-layer` provides a simple `<HeatmapLayer />` component for creating a heatmap layer in a `react-leaflet` map.

![A screenshot of a heatmap on a leaflet map](../master/screenshot.jpg?raw=true)

## Usage

This fork changes the behaviour to export a factory function, so that the types can be turned into a generic

Use directly as a fixed layer:

```typescript
import React from 'react';
import { render } from 'react-dom';
import { Map, Marker, Popup, TileLayer } from 'react-leaflet';
import HeatmapLayerFactory from '../src/HeatmapLayer';
import { addressPoints } from './realworld.10000.js';

const HeatmapLayer = HeatmapLayerFactory<[number, number, number]>()

class MapExample extends React.Component {

  render() {
    return (
      <div>
        <Map center={[0,0]} zoom={13}>
          <HeatmapLayer
            fitBoundsOnLoad
            fitBoundsOnUpdate
            points={addressPoints}
            longitudeExtractor={m => m[1]}
            latitudeExtractor={m => m[0]}
            intensityExtractor={m => parseFloat(m[2])} />
          <TileLayer
            url='http://{s}.tile.osm.org/{z}/{x}/{y}.png'
            attribution='&copy; <a href="http://osm.org/copyright">OpenStreetMap</a> contributors'
          />
        </Map>
      </div>
    );
  }

}

render(<MapExample />, document.getElementById('app'));
```

Or use it inside a layer control to toggle it:

```typescript
import React from 'react';
import { render } from 'react-dom';
import { Map, Marker, Popup, TileLayer } from 'react-leaflet';
import HeatmapLayerFactory from '../src/HeatmapLayer';
import { addressPoints } from './realworld.10000.js';

const HeatmapLayer = HeatmapLayerFactory<[number, number, number]>()

class MapExample extends React.Component {

  render() {
    return (
      <div>
      <Map center={position} zoom={13} style={{ height: '100%' }} >
            <LayersControl>
              <LayersControl.BaseLayer name="Base" checked>
                <TileLayer
                  url="http://{s}.tile.osm.org/{z}/{x}/{y}.png"
                  attribution="&copy; <a href=http://osm.org/copyright>OpenStreetMap</a> contributors"
                />
              </LayersControl.BaseLayer>
              <LayersControl.Overlay name="Heatmap" checked>
                <FeatureGroup color="purple">
                  <Marker position={[50.05, -0.09]} >
                    <Popup>
                      <span>A pretty CSS3 popup.<br /> Easily customizable. </span>
                    </Popup>
                  </Marker>
                  <HeatmapLayer
                    fitBoundsOnLoad
                    fitBoundsOnUpdate
                    points={addressPoints}
                    longitudeExtractor={m => m[1]}
                    latitudeExtractor={m => m[0]}
                    intensityExtractor={m => parseFloat(m[2])}
                  />
                </FeatureGroup>
              </LayersControl.Overlay>
              <LayersControl.Overlay name="Marker">
                <FeatureGroup color="purple">
                  <Marker position={position} >
                    <Popup>
                      <span>A pretty CSS3 popup.<br /> Easily customizable. </span>
                    </Popup>
                  </Marker>
                </FeatureGroup>
              </LayersControl.Overlay>
            </LayersControl>
          </Map>
      </div>
    );
  }
}

render(<MapExample />, document.getElementById('app'));
```


## API

The `HeatmapLayer` component takes the following props:

- `points`: *required* an array of objects to be processed
- `longitudeExtractor`: *required* a function that returns the object's longitude e.g. `marker => marker.lng`
- `latitudeExtractor`: *required* a function that returns the object's latitude e.g. `marker => marker.lat`
- `intensityExtractor`: *required* a function that returns the object's intensity e.g. `marker => marker.val`
- `fitBoundsOnLoad`: boolean indicating whether map should fit data in bounds of map on load
- `fitBoundsOnUpdate`: boolean indicating whether map should fit data in bounds of map on Update
- `max`: max intensity value for heatmap (default: 3.0)
- `radius`: radius for heatmap points (default: 30)
- `maxZoom`: maximum zoom for heatmap (default: 18)
- `opacity`: the opacity of the heatmap layer (default: 1)
- `minOpacity`: minimum opacity for heatmap (default: 0.01)
- `useLocalExtrema`: whether to always have a local minimum and maximum in the viewport (default: false)
- `blur`: blur for heatmap points (default: 15)
- `gradient`: object defining gradient stop points for heatmap e.g. `{ 0.4: 'blue', 0.8: 'orange', 1.0: 'red' }` (default: `simpleheat` package default gradient)
- `onStatsUpdate`: called on redraw with a { min, max } object containing:
  - The local minimum intensity in the viewport
  - The local maximum intensity in the viewport

## Contributing

See [CONTRIBUTING.md](https://www.github.com/OpenGov/react-leaflet-heatmap-layer/blob/master/CONTRIBUTING.md)

## Credits

This package was inspired by [Leaflet.heat](https://github.com/Leaflet/Leaflet.heat).

## License

`react-leaflet-heatmap-layer` is MIT licensed.

See [LICENSE.md](https://www.github.com/OpenGov/react-leaflet-heatmap-layer/blob/master/LICENSE.md) for details.


CHANGELOG.md
# 3.0.0 Release
- react-leaflet v3.x support.  For react-leaflet v2.x please use react-leaflet-heatmep-layer v2.x.

# 2.0.0 Release
- React-leaflet v2.x support. For react-leaflet v1.x please use react-leaflet-heatmep-layer v1.x.

# 1.0.4 Release
- Allow the latest versions of `react` and `react-dom` (i.e. 16 and up).

# 1.0.3 Release
- Fixed warning "Accessing PropTypes via the main React package is deprecated"

# 1.0.2 Release
- Fix bug where radius, blur, and max were not being used when passed in as props.

# 1.0.1 Release
- Fix bug in componentWillUnmount->safeRemoveLayer where getPanes doesn't return anything so .contains is called on undefined. 

# 1.0.0 Release
- Leaflet 1.0.0 support
- React-Leaflet 1.0.0 support
- List eslint as an explicit devDependency
- upgrade the eslint related packages
- fix linting errors using current config
- Add notes about maintaining absence of lint errors and warnings in Contributing guide
- This should make it easier to ensure code quality as others contribute in the open
- Also, drop unused jest and enzyme deps

# 0.2.3 Release
- Missed Transpilation

# 0.2.2 Release
- Change `getHeatmapProps` signature to take a `props` argument to support passing `nextProps` from `componentWillReceiveProps` and `this.props` from `componentDidMount`

# 0.2.1 Release

- Fix getMaxZoom returning props.radius instead of props.maxZoom, fix misnamed call to getMax instead of getMaxZoom in redraw()

# 0.2.0 Release

- adds an `onStatsUpdate` prop which is called on redraw with a { min, max } object containing the min and max number of items found for a single coordinate

# 0.1.0 Release

- Handle invalid points gracefully
- Add `fitBoundsOnUpdate` prop to indicate that the map should fit the data in the map on update

# 0.0.2 Release

- Fix an issue with the gradient property being applied

# 0.0.1 Release

- Initial implementation of package


src/HeatmapLayer.spec.js
import { computeAggregate } from './Heatmap'

const POINTS = [
  {
    coordinates: [1, 2],
    intensity: 5,
  },
  {
    coordinates: [5, 6],
    intensity: 6,
  },
  {
    coordinates: [8, 7],
    intensity: 99,
  },
  {
    coordinates: [9, 1],
    intensity: 1,
  },
]

const AGG = {
  data: {},
  same: new Set(),
  seen: 1,
}

const getKey = (point) => `${point.coordinates[0]} - ${point.coordinates[1]}`

describe('computeAggregate', () => {
  describe('sum', () => {
    test('returns the intensity when given a single intensity', () => {
      const result = computeAggregate(AGG, 5, 'sum')
      expect(result).toBe(5)
    })

    test('returns the intensity for each intensity when given multiple different points', () => {
      const aggregates = {}

      POINTS.forEach((point) => {
        const key = getKey(point)

        aggregates[key] = {
          data: {},
          seen: 0,
        }

        aggregates[key].seen++

        const result = computeAggregate(aggregates[key], point.intensity, 'sum')
        expect(result).toBe(point.intensity)
      })
    })

    test('sums the intensity for each intensity when points overlap', () => {
      const aggregates = {}

      const point = {
        coordinates: [7, 1],
        intensity: 3,
      }

      const key = getKey(point)

      aggregates[key] = {
        data: {},
        seen: 0,
      }

      const n = 20

      for (let i = 0; i < n; i++) {
        aggregates[key].seen++
        computeAggregate(aggregates[key], point.intensity, 'sum')
      }

      expect(aggregates[key].data.sum).toBe(n * point.intensity)
    })
  })

  describe('count', () => {
    test('returns 1 when given a single intensity', () => {
      const result = computeAggregate(AGG, 5, 'count')
      expect(result).toBe(1)
    })

    test('increments the count for the same point each time', () => {
      const aggregates = {}

      const point = {
        coordinates: [5, 5],
        intensity: 3,
      }

      const key = getKey(point)

      aggregates[key] = {
        data: {},
        seen: 0,
      }

      const n = 5

      for (let i = 0; i < n; i++) {
        aggregates[key].seen++
        computeAggregate(aggregates[key], point.intensity, 'count')
      }

      expect(aggregates[key].data.count).toBe(n)
    })
  })

  describe('mean', () => {
    test('returns the intensity when given a single intensity', () => {
      const result = computeAggregate(AGG, 5, 'mean')
      expect(result).toBe(5)
    })

    test('computes a cumulative moving average for each point', () => {
      const aggregates = {}

      const visited = []

      POINTS.forEach((point) => {
        const key = getKey(point)

        if (!aggregates[key]) {
          aggregates[key] = {
            data: {},
            seen: 0,
          }
        }

        aggregates[key].seen++

        const result = computeAggregate(aggregates[key], point.intensity, 'mean')

        // calculate average of all previously visited points
        // and then calculate the new average using a cumulative moving average
        let average = visited.reduce((p, c) => p + c, 0) / visited.length || 0
        average += (point.intensity - average) / aggregates[key].seen

        visited.push(point.intensity)

        expect(result).toBe(average)
      })
    })

    test('computes a final average for all points', () => {
      const aggregates = {}

      const points = [
        {
          coordinates: [5, 7],
          intensity: 4,
        },
        {
          coordinates: [5, 7],
          intensity: 3,
        },
        {
          coordinates: [5, 7],
          intensity: 2,
        },
        {
          coordinates: [5, 7],
          intensity: 9,
        },
        {
          coordinates: [5, 7],
          intensity: 11,
        },
        {
          coordinates: [5, 7],
          intensity: 6,
        },
      ]

      const key = getKey(points[0])

      points.forEach((point) => {
        if (!aggregates[key]) {
          aggregates[key] = {
            data: {},
            same: new Set(),
            seen: 0,
          }
        }

        aggregates[key].seen++

        computeAggregate(aggregates[key], point.intensity, 'mean')
      })

      expect(aggregates[key].data.mean).toBe(5.833333333333333)
    })
  })

  describe('distinct', () => {
    test('returns 1 when given a single unique intensity', () => {
      const result = computeAggregate(AGG, 5, 'distinct')
      expect(result).toBe(1)
    })

    test('returns the number of unique elements when given multiple intensity points', () => {
      const aggregates = {}

      const points = [
        {
          coordinates: [5, 7],
          intensity: 4,
        },
        {
          coordinates: [5, 7],
          intensity: 2,
        },
        {
          coordinates: [5, 7],
          intensity: 9,
        },
      ]

      const key = getKey(points[0])

      points.forEach((point) => {
        if (!aggregates[key]) {
          aggregates[key] = {
            data: {},
            same: new Set(),
            seen: 0,
          }
        }

        aggregates[key].seen++

        computeAggregate(aggregates[key], point.intensity, 'distinct')
      })

      expect(aggregates[key].data.distinct).toBe(3)
    })

    test('returns the number of unique elements when given duplicate intensities', () => {
      const aggregates = {}

      const points = [
        {
          coordinates: [5, 7],
          intensity: 2,
        },
        {
          coordinates: [5, 7],
          intensity: 2,
        },
        {
          coordinates: [5, 7],
          intensity: 9,
        },
        {
          coordinates: [5, 7],
          intensity: 9,
        },
        {
          coordinates: [5, 7],
          intensity: 10,
        },
      ]

      const key = getKey(points[0])

      points.forEach((point) => {
        if (!aggregates[key]) {
          aggregates[key] = {
            data: {},
            same: new Set(),
            seen: 0,
          }
        }

        aggregates[key].seen++

        computeAggregate(aggregates[key], point.intensity, 'distinct')
      })

      expect(aggregates[key].data.distinct).toBe(3)
    })
  })

  describe('min', () => {
    test('returns the intensity when given a single intensity', () => {
      const result = computeAggregate(AGG, 5, 'min')
      expect(result).toBe(5)
    })

    test('returns the min intensity value when given points at the same location', () => {
      const aggregates = {}

      const points = [
        {
          coordinates: [5, 7],
          intensity: 4,
        },
        {
          coordinates: [5, 7],
          intensity: 2,
        },
        {
          coordinates: [5, 7],
          intensity: 9,
        },
      ]

      const key = getKey(points[0])

      points.forEach((point) => {
        if (!aggregates[key]) {
          aggregates[key] = {
            data: {},
            same: new Set(),
            seen: 0,
          }
        }

        aggregates[key].seen++

        computeAggregate(aggregates[key], point.intensity, 'min')
      })

      expect(aggregates[key].data.min).toBe(2)
    })
  })

  describe('max', () => {
    test('returns the intensity when given a single intensity', () => {
      const result = computeAggregate(AGG, 5, 'max')
      expect(result).toBe(5)
    })

    test('returns the max intensity value when given points at the same location', () => {
      const aggregates = {}

      const points = [
        {
          coordinates: [5, 7],
          intensity: 4,
        },
        {
          coordinates: [5, 7],
          intensity: 2,
        },
        {
          coordinates: [5, 7],
          intensity: 9,
        },
      ]

      const key = getKey(points[0])

      points.forEach((point) => {
        if (!aggregates[key]) {
          aggregates[key] = {
            data: {},
            same: new Set(),
            seen: 0,
          }
        }

        aggregates[key].seen++

        computeAggregate(aggregates[key], point.intensity, 'max')
      })

      expect(aggregates[key].data.max).toBe(9)
    })
  })

  describe('stdev', () => {
    test('returns 0 when given a single intensity', () => {
      const result = computeAggregate(AGG, 5, 'stdev')
      expect(result).toBe(0)
    })

    test('returns the stdev when given points at the same location', () => {
      const aggregates = {}

      const points = [
        {
          coordinates: [5, 7],
          intensity: 2,
        },
        {
          coordinates: [5, 7],
          intensity: 5,
        },
        {
          coordinates: [5, 7],
          intensity: 1,
        },
        {
          coordinates: [5, 7],
          intensity: 89,
        },
        {
          coordinates: [5, 7],
          intensity: 6,
        },
      ]

      const key = getKey(points[0])

      points.forEach((point) => {
        if (!aggregates[key]) {
          aggregates[key] = {
            data: {},
            same: new Set(),
            seen: 0,
          }
        }

        aggregates[key].seen++

        computeAggregate(aggregates[key], point.intensity, 'stdev')
      })

      expect(aggregates[key].data.stdev).toBe(38.29229687548136)
    })

    test('returns the stdev when given points at different locations', () => {
      const aggregates = {}

      const points = [
        {
          coordinates: [6, 7],
          intensity: 2,
        },
        {
          coordinates: [9, 2],
          intensity: 5,
        },
        {
          coordinates: [6, 7],
          intensity: 1,
        },
        {
          coordinates: [6, 7],
          intensity: 89,
        },
        {
          coordinates: [1, 4],
          intensity: 6,
        },
      ]

      const targetKey = getKey(points[0])

      points.forEach((point) => {
        const key = getKey(point)

        if (!aggregates[key]) {
          aggregates[key] = {
            data: {},
            same: new Set(),
            seen: 0,
          }
        }

        aggregates[key].seen++

        computeAggregate(aggregates[key], point.intensity, 'stdev')
      })

      expect(aggregates[targetKey].data.stdev).toBe(50.52062285179522)
    })
  })

  describe('stdevp', () => {
    test('returns 0 when given a single intensity', () => {
      const result = computeAggregate(AGG, 5, 'stdevp')
      expect(result).toBe(0)
    })

    test('returns the population stdev when given points at the same location', () => {
      const aggregates = {}

      const points = [
        {
          coordinates: [5, 7],
          intensity: 2,
        },
        {
          coordinates: [5, 7],
          intensity: 5,
        },
        {
          coordinates: [5, 7],
          intensity: 1,
        },
        {
          coordinates: [5, 7],
          intensity: 89,
        },
        {
          coordinates: [5, 7],
          intensity: 6,
        },
      ]

      const key = getKey(points[0])

      points.forEach((point) => {
        if (!aggregates[key]) {
          aggregates[key] = {
            data: {},
            same: new Set(),
            seen: 0,
          }
        }

        aggregates[key].seen++

        computeAggregate(aggregates[key], point.intensity, 'stdevp')
      })

      expect(aggregates[key].data.stdevp).toBe(34.24967153127165)
    })

    test('returns the population stdev when given points at different locations', () => {
      const aggregates = {}

      const points = [
        {
          coordinates: [6, 7],
          intensity: 2,
        },
        {
          coordinates: [9, 2],
          intensity: 5,
        },
        {
          coordinates: [6, 7],
          intensity: 1,
        },
        {
          coordinates: [6, 7],
          intensity: 89,
        },
        {
          coordinates: [1, 4],
          intensity: 6,
        },
      ]

      const targetKey = getKey(points[0])

      points.forEach((point) => {
        const key = getKey(point)

        if (!aggregates[key]) {
          aggregates[key] = {
            data: {},
            same: new Set(),
            seen: 0,
          }
        }

        aggregates[key].seen++

        computeAggregate(aggregates[key], point.intensity, 'stdevp')
      })

      expect(aggregates[targetKey].data.stdevp).toBe(41.24991582482994)
    })
  })

  describe('variance', () => {
    test('returns 0 when given a single intensity', () => {
      const result = computeAggregate(AGG, 5, 'variance')
      expect(result).toBe(0)
    })

    test('returns the sample variance when given points at the same location', () => {
      const aggregates = {}

      const points = [
        {
          coordinates: [5, 7],
          intensity: 2,
        },
        {
          coordinates: [5, 7],
          intensity: 5,
        },
        {
          coordinates: [5, 7],
          intensity: 1,
        },
        {
          coordinates: [5, 7],
          intensity: 89,
        },
        {
          coordinates: [5, 7],
          intensity: 6,
        },
      ]

      const key = getKey(points[0])

      points.forEach((point) => {
        if (!aggregates[key]) {
          aggregates[key] = {
            data: {},
            same: new Set(),
            seen: 0,
          }
        }

        aggregates[key].seen++

        computeAggregate(aggregates[key], point.intensity, 'variance')
      })

      expect(aggregates[key].data.variance).toBe(1466.3)
    })

    test('returns the sample variance when given points at different locations', () => {
      const aggregates = {}

      const points = [
        {
          coordinates: [6, 7],
          intensity: 2,
        },
        {
          coordinates: [9, 2],
          intensity: 5,
        },
        {
          coordinates: [6, 7],
          intensity: 1,
        },
        {
          coordinates: [6, 7],
          intensity: 89,
        },
        {
          coordinates: [1, 4],
          intensity: 6,
        },
      ]

      const targetKey = getKey(points[0])

      points.forEach((point) => {
        const key = getKey(point)

        if (!aggregates[key]) {
          aggregates[key] = {
            data: {},
            same: new Set(),
            seen: 0,
          }
        }

        aggregates[key].seen++

        computeAggregate(aggregates[key], point.intensity, 'variance')
      })

      expect(aggregates[targetKey].data.variance).toBe(2552.333333333333)
    })
  })

  describe('variancep', () => {
    test('returns 0 when given a single intensity', () => {
      const result = computeAggregate(AGG, 5, 'variancep')
      expect(result).toBe(0)
    })

    test('returns the population variance when given points at the same location', () => {
      const aggregates = {}

      const points = [
        {
          coordinates: [5, 7],
          intensity: 2,
        },
        {
          coordinates: [5, 7],
          intensity: 5,
        },
        {
          coordinates: [5, 7],
          intensity: 1,
        },
        {
          coordinates: [5, 7],
          intensity: 89,
        },
        {
          coordinates: [5, 7],
          intensity: 6,
        },
      ]

      const key = getKey(points[0])

      points.forEach((point) => {
        if (!aggregates[key]) {
          aggregates[key] = {
            data: {},
            same: new Set(),
            seen: 0,
          }
        }

        aggregates[key].seen++

        computeAggregate(aggregates[key], point.intensity, 'variancep')
      })

      expect(aggregates[key].data.variancep).toBe(1173.04)
    })

    test('returns the population variance when given points at different locations', () => {
      const aggregates = {}

      const points = [
        {
          coordinates: [6, 7],
          intensity: 2,
        },
        {
          coordinates: [9, 2],
          intensity: 5,
        },
        {
          coordinates: [6, 7],
          intensity: 1,
        },
        {
          coordinates: [6, 7],
          intensity: 89,
        },
        {
          coordinates: [1, 4],
          intensity: 6,
        },
      ]

      const targetKey = getKey(points[0])

      points.forEach((point) => {
        const key = getKey(point)

        if (!aggregates[key]) {
          aggregates[key] = {
            data: {},
            same: new Set(),
            seen: 0,
          }
        }

        aggregates[key].seen++

        computeAggregate(aggregates[key], point.intensity, 'variancep')
      })

      expect(aggregates[targetKey].data.variancep).toBe(1701.5555555555554)
    })
  })
})


src/Heatmap.ts
import * as L from 'leaflet'
/// <reference path="simpleheat.d.ts"/>
import SimpleHeat, { Gradient } from '@vgrid/simpleheat'

export { SimpleHeat, Gradient }

export type AggregateType =
  | 'mean'
  | 'count'
  | 'sum'
  | 'distinct'
  | 'min'
  | 'max'
  | 'variance'
  | 'variancep'
  | 'stdev'
  | 'stdevp'

export type Aggregation = {
  data: {
    mean: number
    sigma: number
    count: number
    sum: number
    distinct: number
    min: number
    max: number
    variance: number
    variancep: number
    stdev: number
    stdevp: number
  }
  same: Set<number>
  seen: number
}

const max = (arr: number[]) =>
  arr.reduce((_max, value) => Math.max(_max, value), Number.MIN_SAFE_INTEGER)
const min = (arr: number[]) =>
  arr.reduce((_min, value) => Math.min(_min, value), Number.MAX_SAFE_INTEGER)

function isNumber(val: unknown): val is number {
  return typeof val === 'number'
}

function isInvalid(num: number): boolean {
  return !isNumber(num) && !num
}

function isValid(num: number): boolean {
  return !isInvalid(num)
}

function isValidLatLngArray(arr: Array<number>): boolean {
  return arr.filter(isValid).length === arr.length
}

function isInvalidLatLngArray(arr: Array<number>): boolean {
  return !isValidLatLngArray(arr)
}

function shouldIgnoreLocation(loc: { lat: number; lng: number }): boolean {
  return isInvalid(loc.lng) || isInvalid(loc.lat)
}

export function computeAggregate(
  agg: Aggregation,
  intensity: number,
  aggregateType: AggregateType = 'sum',
): number {
  const updateMeanAndSigma = (c: number, v: number) => {
    const newMean = agg.data.mean + fns.mean(agg.data.mean, c, v)
    agg.data.sigma += (v - newMean) * (v - agg.data.mean)
    agg.data.mean = newMean
  }

  const fns = {
    mean: (m: number, c: number, v: number) => (v - m) / c,
    count: () => 1,
    sum: (m: number, c: number, v: number) => v,
    distinct: (m: number, c: number, v: number) => {
      agg.same.add(v)
      return agg.same.size
    },
    min: (m: number, c: number, v: number) => Math.min(m, v),
    max: (m: number, c: number, v: number) => Math.max(m, v),
    variance: (m: number, c: number, v: number) => {
      updateMeanAndSigma(c, v)
      return c > 1 ? agg.data.sigma / (c - 1) : 0
    },
    variancep: (m: number, c: number, v: number) => {
      updateMeanAndSigma(c, v)
      return c > 1 ? agg.data.sigma / c : 0
    },
    stdev: (m: number, c: number, v: number) => Math.sqrt(fns.variance(m, c, v)),
    stdevp: (m: number, c: number, v: number) => Math.sqrt(fns.variancep(m, c, v)),
  }

  const type = aggregateType.toLowerCase() as AggregateType

  if (!agg.data[type]) {
    if (type === 'min') {
      agg.data[type] = Number.MAX_SAFE_INTEGER
    } else if (type === 'max') {
      agg.data[type] = Number.MIN_SAFE_INTEGER
    } else if (['stdev', 'stdevp', 'variance', 'variancep'].includes(type)) {
      if (!agg.data.mean) {
        agg.data.mean = 0
      }
      if (!agg.data.sigma) {
        agg.data.sigma = 0
      }
      agg.data[type] = 0
    } else {
      agg.data[type] = 0
    }
  }

  const res = (fns[type] || fns.sum)(agg.data[type], agg.seen, intensity)

  if (['mean', 'count', 'sum'].includes(type)) {
    agg.data[type] += res
  } else {
    agg.data[type] = res
  }

  return agg.data[type]
}

export interface SimpleHeatOptions {
  opacity: number
  minOpacity: number
  maxZoom: number
  radius: number
  blur: number
  max: number
  gradient?: Gradient
}

export interface HeatmapOptions<Point> extends L.LayerOptions, Partial<SimpleHeatOptions> {
  points: Point[]
  longitudeExtractor: (point: Point) => number
  latitudeExtractor: (point: Point) => number
  intensityExtractor: (point: Point) => number
  fitBoundsOnLoad?: boolean
  fitBoundsOnUpdate?: boolean
  onStatsUpdate?: (stats: { min: number; max: number }) => void

  useLocalExtrema?: boolean
  aggregateType?: AggregateType
}

type Cell = [number, number, number]

export default class Heatmap<Point> extends L.Layer {
  private __el?: HTMLCanvasElement
  private __heatmap?: SimpleHeat
  private _frame?: number | null
  options: HeatmapOptions<Point>

  constructor(options?: HeatmapOptions<Point>) {
    super(options)
    this.options = L.Util.setOptions(this, options)
  }

  private get _heatmap() {
    if (!this.__heatmap) {
      this.__el = document.createElement('canvas')
      this.__heatmap = new SimpleHeat(this.__el)
    }
    return this.__heatmap
  }

  private get _el() {
    if (!this.__el) {
      this.__el = document.createElement('canvas')
      this.__heatmap = new SimpleHeat(this.__el)
    }
    return this.__el
  }

  getPane(): HTMLElement {
    return super.getPane() ?? this._map.getPanes().overlayPane
  }

  onAdd(map: L.Map): this {
    const canAnimate = map.options.zoomAnimation && L.Browser.any3d
    const zoomClass = `leaflet-zoom-${canAnimate ? 'animated' : 'hide'}`
    const mapSize = map.getSize()

    this._el.className = zoomClass
    this._el.style.transformOrigin = '50% 50%'
    this._el.width = mapSize.x
    this._el.height = mapSize.y
    this._heatmap.resize()

    this.getPane().appendChild(this._el)

    this.reset()

    if (this.options.fitBoundsOnLoad) {
      this.fitBounds()
    }

    this.updateSimpleHeat(this.getSimpleHeatOptions())

    return this
  }

  onRemove(): this {
    const pane = this.getPane()
    if (pane.contains(this._el)) {
      pane.removeChild(this._el)
    }
    return this
  }

  getEvents(): { [name: string]: L.LeafletEventHandlerFn } {
    return {
      viewreset: this.reset,
      moveend: this.reset,
      zoomanim: this._animateZoom,
    }
  }

  _animateZoom(e: L.LeafletEvent): void {
    const _e = e as L.ZoomAnimEvent

    const scale = this._map.getZoomScale(_e.zoom)
    const offset = this._map
      .latLngToLayerPoint(_e.center)
      .subtract(this._map.containerPointToLayerPoint(this._map.getSize().divideBy(2)))
      .multiplyBy(-scale)
      .subtract(this._map.layerPointToContainerPoint([0, 0]))

    L.DomUtil.setTransform(this._el, offset, scale)
  }

  fitBounds(): void {
    const { points, longitudeExtractor, latitudeExtractor } = this.options

    const lngs = points.map(longitudeExtractor)
    const lats = points.map(latitudeExtractor)

    const ne = { lng: max(lngs), lat: max(lats) }
    const sw = { lng: min(lngs), lat: min(lats) }

    if (shouldIgnoreLocation(ne) || shouldIgnoreLocation(sw)) {
      return
    }

    this._map.fitBounds(L.latLngBounds(L.latLng(sw), L.latLng(ne)))
  }

  resize(): void {
    if (!this._map) {
      return
    }

    const size = this._map.getSize()
    if (size.x !== this._el.width || size.y !== this._el.height) {
      this._el.width = size.x
      this._el.height = size.y
      this._heatmap.resize()
    }
  }

  getMinOpacity(): number {
    return this.options.minOpacity ?? 0.01
  }

  getOpacity(): number {
    return this.options.opacity ?? 1
  }

  getMax(): number {
    return this.options.max ?? 3.0
  }

  getRadius(): number {
    return this.options.radius ?? 30
  }

  getMaxZoom(): number {
    return this.options.maxZoom ?? 18
  }

  getBlur(): number {
    return this.options.blur ?? 15
  }

  getSimpleHeatOptions(): SimpleHeatOptions {
    return {
      opacity: this.getOpacity(),
      minOpacity: this.getMinOpacity(),
      maxZoom: this.getMaxZoom(),
      radius: this.getRadius(),
      blur: this.getBlur(),
      max: this.getMax(),
      gradient: this.options.gradient,
    }
  }

  /**
   * Update various heatmap properties like radius, gradient, and max
   */
  updateSimpleHeat(options: Partial<SimpleHeatOptions>): void {
    this.updateHeatmapRadius(options.radius, options.blur)
    this.updateHeatmapGradient(options.gradient)
    this.updateHeatmapMax(options.max)
  }

  /**
   * Update the heatmap's radius and blur (blur is optional)
   */
  updateHeatmapRadius(radius?: number, blur?: number): void {
    if (isNumber(radius)) {
      this._heatmap.radius(radius, blur)
    }
  }

  /**
   * Update the heatmap's gradient
   */
  updateHeatmapGradient(gradient?: Gradient): void {
    if (gradient) {
      this._heatmap.gradient(gradient)
    }
  }

  /**
   * Update the heatmap's maximum
   */
  updateHeatmapMax(maximum?: number): void {
    if (isNumber(maximum)) {
      this._heatmap.max(maximum)
    }
  }

  redraw(): void {
    if (!this._map) {
      return
    }

    const r = this._heatmap._r
    const size = this._map.getSize()
    const cellSize = r / 2
    const panePos = this._map.layerPointToContainerPoint([0, 0])
    const offsetX = panePos.x % cellSize
    const offsetY = panePos.y % cellSize

    const getLat = this.options.latitudeExtractor
    const getLng = this.options.longitudeExtractor
    const getIntensity = this.options.intensityExtractor
    const inBounds = (p: L.PointExpression, bounds: L.Bounds) => bounds.contains(p)

    const filterUndefined = (row: Cell[]) => row.filter((c) => c !== undefined)
    const roundResults = (results: Cell[][]) =>
      results.reduce(
        (result, row) =>
          filterUndefined(row)
            .map((cell) => [Math.round(cell[0]), Math.round(cell[1]), cell[2]] as Cell)
            .concat(result),
        [],
      )

    const aggregates: { [key: string]: Aggregation } = {}

    const accumulateInGrid = (
      points: Point[],
      leafletMap: L.Map,
      bounds: L.Bounds,
      aggregateType?: AggregateType,
    ) =>
      points.reduce((grid, point) => {
        const latLng: [number, number] = [getLat(point), getLng(point)]

        //skip invalid points
        if (isInvalidLatLngArray(latLng)) {
          return grid
        }

        const p = leafletMap.latLngToContainerPoint(latLng)

        if (!inBounds(p, bounds)) {
          return grid
        }

        const x = Math.floor((p.x - offsetX) / cellSize) + 2
        const y = Math.floor((p.y - offsetY) / cellSize) + 2

        grid[y] = grid[y] || []
        const cell = grid[y][x]
        const key = `${x}-${y}`

        if (!aggregates[key]) {
          aggregates[key] = {
            data: {} as Aggregation['data'],
            same: new Set(),
            seen: 0,
          }
        }

        aggregates[key].seen++

        const intensity = getIntensity(point)
        const agg = computeAggregate(aggregates[key], intensity, aggregateType)

        if (!cell) {
          grid[y][x] = [p.x, p.y, agg]
        } else {
          cell[0] = (cell[0] * cell[2] + p.x * agg) / (cell[2] + agg) // x
          cell[1] = (cell[1] * cell[2] + p.y * agg) / (cell[2] + agg) // y
          cell[2] = agg
        }

        return grid
      }, [] as Cell[][])

    const getBounds = () => new L.Bounds(L.point([-r, -r]), size.add([r, r]))

    const getDataForHeatmap = (points: Point[], leafletMap: L.Map, aggregateType?: AggregateType) =>
      roundResults(accumulateInGrid(points, leafletMap, getBounds(/*leafletMap*/), aggregateType))

    const data = getDataForHeatmap(this.options.points, this._map, this.options.aggregateType)

    const totalMax = max(data.map((m) => m[2]))

    this._heatmap.clear()
    this._heatmap.data(data)

    if (this.options.useLocalExtrema) {
      this.updateHeatmapMax(totalMax)
    }

    try {
      this._heatmap.draw(this.getMinOpacity())
    } catch (DOMException) {
      // Safe to ignore - occurs if the width or height is 0
    }

    this._frame = null
    if (this.options.onStatsUpdate && this.options.points && this.options.points.length > 0) {
      this.options.onStatsUpdate({
        min: min(data.map((m) => m[2])),
        max: totalMax,
      })
    }

    this._el.style.opacity = this.getOpacity().toString()
  }

  reset(): void {
    if (!this._map) {
      return
    }

    const topLeft = this._map.containerPointToLayerPoint([0, 0])
    L.DomUtil.setPosition(this._el, topLeft)

    this.resize()

    if (this._heatmap && !this._frame /*&& !this._map._animating*/) {
      this._frame = L.Util.requestAnimFrame(this.redraw, this)
    }

    this.redraw()
  }
}
